### PR TITLE
FIR: Implement checker on open members

### PR DESCRIPTION
**Pull request by tgeng**
Implement the following checks:

1. NON_FINAL_MEMBER_IN_FINAL_CLASS
2. NON_FINAL_MEMBER_IN_OBJECT

### CODE REVIEW

File: compiler/testData/diagnostics/tests/modifiers/IllegalModifiers.fir.kt [hunk 1]
```diff
@@ -7,14 +7,14 @@ abstract class A() {
     <!INCOMPATIBLE_MODIFIERS!>final<!> <!INCOMPATIBLE_MODIFIERS!>open<!> fun h() {}
 
     open var r: String
-    get
-    abstract protected set
+        get
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/modifiers/IllegalModifiers.fir.kt:10]: There are many unnecessary changes regarding spaces/intents. Are those test comparison-generated or accidentally added?

File: compiler/fir/entrypoint/src/org/jetbrains/kotlin/fir/checkers/CommonDeclarationCheckers.kt [hunk 2]
```diff
@@ -58,6 +58,7 @@ object CommonDeclarationCheckers : DeclarationCheckers() {
         FirTypeParametersInObjectChecker,
         FirMemberFunctionChecker,
         FirMemberPropertyChecker,
+        FirOpenMemberChecker,
```
Comment (jsjeon) [inline compiler/fir/entrypoint/src/org/jetbrains/kotlin/fir/checkers/CommonDeclarationCheckers.kt:61]: `NON_FINAL_MEMBER_IN_OBJECT` indicates this diagnostic wants to visit and check `object`? Then, this should be in `classCheckers`, not `regularClassChecker`, which doesn't visit anonymous objects.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirCallableMemberDeclaration
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.modality
+
+object FirOpenMemberChecker : FirRegularClassChecker() {
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt:18]: Same here. You need to extend `FirClassChecker` instead, in order to visit `declaration: FirClass<*>`.

File: compiler/testData/diagnostics/tests/modifiers/IllegalModifiers.fir.kt (hunk 1 above)
Comment (tgeng) [reply compiler/testData/diagnostics/tests/modifiers/IllegalModifiers.fir.kt:10]: It's generated by the test infra.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    override fun check(declaration: FirRegularClass, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.canHaveOpenMembers()) return
+        for (memberDeclaration in declaration.declarations) {
+            if (memberDeclaration !is FirCallableMemberDeclaration<*>) continue
+            val source = memberDeclaration.source
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt:23]: val source = memberDeclaration.source ?: continue

because we don't report on `null` source.

Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt:23]: One more thing to check here. We don't want to report on _fake_ source, either.

File: compiler/fir/entrypoint/src/org/jetbrains/kotlin/fir/checkers/CommonDeclarationCheckers.kt (hunk 2 above)
Comment (tgeng) [reply compiler/fir/entrypoint/src/org/jetbrains/kotlin/fir/checkers/CommonDeclarationCheckers.kt:61]: Thanks! Fixed

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt (hunk 4 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt]: If we don't report fake, the following `x` will be missed

```
value class A8(open val x: Int)
```

Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt]: Yeah, as discussed offline, let's allow fake source for value parameter of `inline`/`value` classes.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import org.jetbrains.kotlin.fir.declarations.FirClass
+import org.jetbrains.kotlin.fir.declarations.modality
+
+object FirOpenMemberChecker : FirClassChecker() {
+    override fun check(declaration: FirClass<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        if (declaration.canHaveOpenMembers()) return
+        for (memberDeclaration in declaration.declarations) {
+            if (memberDeclaration !is FirCallableMemberDeclaration<*>) continue
+            val source = memberDeclaration.source ?: continue
+            if (memberDeclaration.modality == Modality.OPEN) {
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt:24]: Sorry for not catching this earlier. If a certain declaration has many, conflicting modalities, it's not guaranteed that checking this is enough to pick up `open` members. So you need to something like this:
```kt
// If multiple (potentially conflicting) modality modifiers are specified, not all modifiers are recorded at `status`.
// So, our source of truth should be the full modifier list retrieved from the source.
val modifierList = with(FirModifierList) { source.getModifierList() }
val hasOpenModifier = modifierList?.modifiers?.any { it.token == KtTokens.OPEN_KEYWORD } == true
val isOpen = memberDelcaration.isOpen || hasOpenModifier
```

Btw, you can use `FirMemberDeclaration.isOpen` which does that modality check.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt [hunk 6]
```diff
@@ -123,3 +126,6 @@ internal fun checkPropertyInitializer(
 private val FirProperty.hasAccessorImplementation: Boolean
     get() = (getter !is FirDefaultPropertyAccessor && getter?.hasBody == true) ||
             (setter !is FirDefaultPropertyAccessor && setter?.hasBody == true)
+
+
+internal fun FirClass<*>.canHaveOpenMembers(): Boolean = modality() != Modality.FINAL || classKind == ClassKind.ENUM_CLASS
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDeclarationCheckerUtils.kt:131]: Nit: This could be `val` since no parameter/context is needed.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt (hunk 4 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt]: But, due to the modifier list from source (below), it's better to filter out some fake sources here.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt (hunk 5 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirOpenMemberChecker.kt]: OK, let me filter on the type of fake source element kind.

**Review by jsjeon — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍
